pkg/controllers/cloud/cloudprovideraccount_controller.go

```diff
@@ -215,25 +245,107 @@ func (r *CloudProviderAccountReconciler) startPollingThread(namespacedName *type
 
 func (r *CloudProviderAccountReconciler) addAccountProviderType(namespacedName *types.NamespacedName,
 	provider runtimev1alpha1.CloudProvider) common.ProviderType {
-	r.mutex.Lock()
-	defer r.mutex.Unlock()
-
 	providerType := common.ProviderType(provider)
 	r.accountProviderType[*namespacedName] = providerType
-
 	return providerType
 }
 
 func (r *CloudProviderAccountReconciler) removeAccountProviderType(namespacedName *types.NamespacedName) {
-	r.mutex.Lock()
-	defer r.mutex.Unlock()
-
 	delete(r.accountProviderType, *namespacedName)
 }
 
 func (r *CloudProviderAccountReconciler) getAccountProviderType(namespacedName *types.NamespacedName) common.ProviderType {
-	r.mutex.Lock()
-	defer r.mutex.Unlock()
-
 	return r.accountProviderType[*namespacedName]
 }
+
+// getCpaBySecret returns nil only when the Secret is not used by any CloudProvideAccount CR,
+// otherwise the dependent CloudProvideAccount CRs will be returned.
+func (r *CloudProviderAccountReconciler) getCpaBySecret(s types.NamespacedName) (error, []crdv1alpha1.CloudProviderAccount) {
+	cpaList := &crdv1alpha1.CloudProviderAccountList{}
+	if err := r.Client.List(context.TODO(), cpaList, &client.ListOptions{}); err != nil {
+		return fmt.Errorf("failed to get CloudProviderAccount list, err %v", err), nil
+	}
+	var cpaItems []crdv1alpha1.CloudProviderAccount
+	for _, cpa := range cpaList.Items {
+		if cpa.Spec.AWSConfig != nil {
+			if cpa.Spec.AWSConfig.SecretRef.Name == s.Name &&
+				cpa.Spec.AWSConfig.SecretRef.Namespace == s.Namespace {
+				cpaItems = append(cpaItems, cpa)
+			}
+		}
+
+		if cpa.Spec.AzureConfig != nil {
+			if cpa.Spec.AzureConfig.SecretRef.Name == s.Name &&
+				cpa.Spec.AzureConfig.SecretRef.Namespace == s.Namespace {
+				cpaItems = append(cpaItems, cpa)
+			}
+		}
+	}
+	return nil, cpaItems
+}
+
+// watchSecret watch the Secret objects.
+func (r *CloudProviderAccountReconciler) watchSecret() {
+	for {
+		event, ok := <-r.watcher.ResultChan()
+		if !ok {
+			r.resetSecretWatcher()
+		} else {
+			switch event.Type {
+			case watch.Modified:
+				r.mutex.Lock()
+				secret := event.Object.(*v1.Secret)
+				namespacedName := types.NamespacedName{Namespace: secret.Namespace, Name: secret.Name}
+				r.Log.WithName("Secret").Info("Received request", "Secret", namespacedName, "operation", "update")
```

Comment:
Instead of prefixing all the log messages with r.Log.WithName("Secret")
can we do initialize a local variable and use it?
log :=  Log.WithName("Secret")

Reply:
Then we should initialize logger for each secret methods.

Reply:
In the other functions, you can keep r.Log.WithName("Secret")

Reply:
Why only this function requires variable

Reply:
Solved.

Reply:
Is it temporary, or this is the way we will have Secret prefix ?